---
title: "Differential_Gene_Expression"
author: "Judy Liang"
date: "2025-01-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set CRAN mirror
options(repos = c(CRAN = "https://cloud.r-project.org/"))
#install.packages("variancePartition", repos='http://cran.us.r-project.org')
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

#BiocManager::install("DESeq2")

# Load necessary packages
library(ggplot2)
library(readxl)
library(tidyverse)
library(janitor)
library(DESeq2)

#load necessary data
load("~/Downloads/april_5_gene_matrix.RData")
load("ALS_combined_cb_fc.RData")

```


```{r}
#processing the data containing all the sample IDs
#make sure all the columns we are interested are cleaned up (right type, no NAs)

ALS_combined_cb_fc$post_mortem_interval_in_hours[is.na(ALS_combined_cb_fc$post_mortem_interval_in_hours)] <- median(ALS_combined_cb_fc$post_mortem_interval_in_hours, na.rm = TRUE)

  
ALS_combined_cb_fc$age_at_death[is.na(ALS_combined_cb_fc$age_at_death)] <- median(ALS_combined_cb_fc$age_at_death, na.rm = TRUE)


ALS_combined_cb_fc$rin[is.na(ALS_combined_cb_fc$rin)] <- median(ALS_combined_cb_fc$rin, na.rm = TRUE)

#clinical factors examined: external_sample_id, rin, age_at_death, post_mortem_interval_in_hours, prep, sex, site_specimen_collected, disease_short

#ensure categorical factors are factorized ()
#scale(as.numeric(metadata_fc1$age_at_death), center= TRUE)
ALS_combined_cb_fc$disease_short <- as.factor(ALS_combined_cb_fc$disease_short)
ALS_combined_cb_fc$sex <- as.factor(ALS_combined_cb_fc$sex)
ALS_combined_cb_fc$prep <- as.factor(ALS_combined_cb_fc$prep)
ALS_combined_cb_fc$site_specimen_collected <- as.factor(ALS_combined_cb_fc$site_specimen_collected)

# read in Picard tools file 
picard_tools <- read.table("/Users/judyliang/Downloads/multiqc_picard_RnaSeqMetrics.txt", header = TRUE, sep = "\t")

#combine ALS_filtered_combined_cortex samples with technical variables 
samples_data <- ALS_combined_cb_fc %>%
  select(external_sample_id, sample_source, rin, age_at_death, post_mortem_interval_in_hours, prep, sex, site_specimen_collected, disease_short) %>%
  inner_join(picard_tools, by = c("external_sample_id" = "Sample"))

#removing "other" and "ALS/other" group from disease_short
samples_data <- samples_data %>%
  filter(!disease_short %in% c("Other", "ALS/other"))

#display column names of samples_data that contains only NA values 
na_columns <- colnames(samples_data)[apply(samples_data, 2, function(x) all(is.na(x)))]
print(na_columns)
#remove columns 
zero_columns <- colnames(samples_data)[apply(samples_data, 2, function(x) all(x == 0, na.rm = TRUE))]
print(zero_columns)

samples_data <- samples_data %>% select(-c(na_columns, zero_columns))

#samples_data <- samples_data %>% select(-c(SAMPLE, LIBRARY, READ_GROUP, IGNORED_READS, CORRECT_STRAND_READS, INCORRECT_STRAND_READS, PCT_CORRECT_STRAND_READS))

```

```{r}
#make sure the gene counts are raw and the form is a matrix 
geneCountsRaw <- as.matrix(genes_counts)

#removing lowly expressed genes
low_expressed_genes <-rowMeans(genes_tpm) < 1
geneCountsRaw <- geneCountsRaw[!low_expressed_genes, ] 

#only keep matching samples between the ALS dataset sample IDs and the sample IDs on the gene counts dataset
sample_ids <- intersect(samples_data$external_sample_id, colnames(geneCountsRaw))
geneCountsRaw <- geneCountsRaw[, sample_ids, drop = FALSE]

samples_data <- samples_data[samples_data$external_sample_id %in% sample_ids, ]

#order the sample_ids to match column orders in genes dataset
samples_data <- samples_data[match(colnames(geneCountsRaw), samples_data$external_sample_id), ]

#round values in geneCountsRaw to make sure counts are all integers
geneCountsRaw <- round(geneCountsRaw)
```

```{r}
#filter out only frontal cortex samples from metadata
fc_samples <- samples_data %>% 
  filter(samples_data$sample_source == "Cortex_Frontal") %>%
  select(-c(sample_source))

fcCounts <- geneCountsRaw[, fc_samples$external_sample_id]

#preparing DESeq dataset
dds_fc <- DESeqDataSetFromMatrix(countData = fcCounts,
                              colData = fc_samples,
                              design = ~ disease_short)

#run differential analysis
dds_fc <- DESeq(dds_fc)

#Extract results to compare each disease group to control
fc_res_als_vs_control <- results(dds_fc, contrast = c("disease_short", "ALS", "Control"))
fc_res_ftd_vs_control <- results(dds_fc, contrast = c("disease_short", "FTD", "Control"))
fc_res_als_ftd_vs_control <- results(dds_fc, contrast = c("disease_short", "ALS/FTD", "Control"))
#fc_res_als_other_vs_control <- results(dds_fc, contrast = c("disease_short", "ALS/other", "Control"))
#fc_res_other_vs_control <- results(dds_fc, contrast = c("disease_short", "Other", "Control"))
fc_res_als_ad_vs_control <- results(dds_fc, contrast = c("disease_short", "ALS/AD", "Control"))


#Control vs ALS
fc_res_als_vs_control <- fc_res_als_vs_control[!is.na(fc_res_als_vs_control$padj), ]
fc_res_als_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```

```{r}
#Control vs FTD
fc_res_ftd_vs_control <- fc_res_ftd_vs_control[!is.na(fc_res_ftd_vs_control$padj), ]
fc_res_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs ALS/FTD
fc_res_als_ftd_vs_control<- fc_res_als_ftd_vs_control[!is.na(fc_res_als_ftd_vs_control$padj), ]

fc_res_als_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```


```{r}
# # Control vs ALS/other
# fc_res_als_other_vs_control<- fc_res_als_other_vs_control[!is.na(fc_res_als_other_vs_control$padj), ]
# 
# fc_res_als_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
# #Control vs Other
# fc_res_other_vs_control<- fc_res_other_vs_control[!is.na(fc_res_other_vs_control$padj), ]
# fc_res_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05)) 
```

```{r}
#Control vs ALS/AD
fc_res_als_ad_vs_control<- fc_res_als_ad_vs_control[!is.na(fc_res_als_ad_vs_control$padj), ]

fc_res_als_ad_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05)) +
  theme(legend.position = "none")
```


```{r}
#fc summary table 
fc_degs_als_vs_control <- fc_res_als_vs_control[which(fc_res_als_vs_control$padj < 0.05), ]
fc_degs_ftd_vs_control <- fc_res_ftd_vs_control[which(fc_res_ftd_vs_control$padj < 0.05), ]
fc_degs_als_ftd_vs_control <- fc_res_als_ftd_vs_control[which(fc_res_als_ftd_vs_control$padj < 0.05), ]
#fc_degs_als_other_vs_control <- fc_res_als_other_vs_control[which(fc_res_als_other_vs_control$padj < 0.05), ]
#fc_degs_other_vs_control <- fc_res_other_vs_control[which(fc_res_other_vs_control$padj < 0.05), ]
fc_degs_als_ad_vs_control <- fc_res_als_ad_vs_control[which(fc_res_als_ad_vs_control$padj < 0.05), ]

# summary_table_fc <- data.frame(
#   Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/other vs Control", "Other vs Control", "ALS/AD vs Control"),
  # Number_of_DEGs = c(nrow(fc_degs_als_vs_control), nrow(fc_degs_ftd_vs_control), nrow(fc_degs_als_ftd_vs_control), nrow(fc_degs_als_other_vs_control), nrow(fc_degs_other_vs_control), nrow(fc_degs_als_ad_vs_control)))

summary_table_fc <- data.frame(
  Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/AD vs Control"),
Number_of_DEGs = c(nrow(fc_degs_als_vs_control), nrow(fc_degs_ftd_vs_control), nrow(fc_degs_als_ftd_vs_control), nrow(fc_degs_als_ad_vs_control)))
summary_table_fc

```

```{r}
#filter out only cerebellum samples from metadata
cb_samples <- samples_data%>% 
  filter(samples_data$sample_source == "Cerebellum") %>%
  select(-c(sample_source))

#filter out columns in geneCountsRaw to only include fc samples
cbCounts <- geneCountsRaw[, cb_samples$external_sample_id]

#preparing DESeq dataset
dds_cb <- DESeqDataSetFromMatrix(countData = cbCounts,
                              colData = cb_samples,
                              design = ~ disease_short)

#run differential analysis
dds_cb <- DESeq(dds_cb)

#Extract results to compare each disease group to control
cb_res_als_vs_control <- results(dds_cb, contrast = c("disease_short", "ALS", "Control"))
cb_res_ftd_vs_control <- results(dds_cb, contrast = c("disease_short", "FTD", "Control"))
cb_res_als_ftd_vs_control <- results(dds_cb, contrast = c("disease_short", "ALS/FTD", "Control"))
#cb_res_als_other_vs_control <- results(dds_cb, contrast = c("disease_short", "ALS/other", "Control"))
#cb_res_other_vs_control <- results(dds_cb, contrast = c("disease_short", "Other", "Control"))
cb_res_als_ad_vs_control <- results(dds_cb, contrast = c("disease_short", "ALS/AD", "Control"))


#Control vs ALS (cb)
cb_res_als_vs_control<- cb_res_als_vs_control[!is.na(cb_res_als_vs_control$padj), ]
cb_res_als_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```

```{r}
#Control vs FTD (cb)
cb_res_ftd_vs_control<- cb_res_ftd_vs_control[!is.na(cb_res_ftd_vs_control$padj), ]
cb_res_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```


```{r}
#Control vs ALS/FTD (cb)
cb_res_als_ftd_vs_control<- cb_res_als_ftd_vs_control[!is.na(cb_res_als_ftd_vs_control$padj), ]
cb_res_als_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```


```{r}
# # Control vs ALS/other (cb)
# cb_res_als_other_vs_control<- cb_res_als_other_vs_control[!is.na(cb_res_als_other_vs_control$padj), ]
# cb_res_als_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
# #Control vs Other (cb)
# cb_res_other_vs_control<- cb_res_other_vs_control[!is.na(cb_res_other_vs_control$padj), ]
# cb_res_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs ALS/AD (cb)
cb_res_als_ad_vs_control<- cb_res_als_ad_vs_control[!is.na(cb_res_als_ad_vs_control$padj), ]
cb_res_als_ad_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```

```{r}
#cb summary table
cb_degs_als_vs_control <- cb_res_als_vs_control[which(cb_res_als_vs_control$padj < 0.05), ]
cb_degs_ftd_vs_control <- cb_res_ftd_vs_control[which(cb_res_ftd_vs_control$padj < 0.05), ]
cb_degs_als_ftd_vs_control <- cb_res_als_ftd_vs_control[which(cb_res_als_ftd_vs_control$padj < 0.05), ]
#cb_degs_als_other_vs_control <- cb_res_als_other_vs_control[which(cb_res_als_other_vs_control$padj < 0.05), ]
#cb_degs_other_vs_control <- cb_res_other_vs_control[which(cb_res_other_vs_control$padj < 0.05), ]
cb_degs_als_ad_vs_control <- cb_res_als_ad_vs_control[which(cb_res_als_ad_vs_control$padj < 0.05), ]

# summary_table_cb <- data.frame(
#   Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/other vs Control", "Other vs Control", "ALS/AD vs Control"),
#   Number_of_DEGs = c(nrow(cb_degs_als_vs_control), nrow(cb_degs_ftd_vs_control), nrow(cb_degs_als_ftd_vs_control), nrow(cb_degs_als_other_vs_control), nrow(cb_degs_other_vs_control), nrow(cb_degs_als_ad_vs_control)))
summary_table_cb <- data.frame(
  Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/AD vs Control"),
  Number_of_DEGs = c(nrow(cb_degs_als_vs_control), nrow(cb_degs_ftd_vs_control), nrow(cb_degs_als_ftd_vs_control), nrow(cb_degs_als_ad_vs_control)))
summary_table_cb

```



```{r}
#creating second model for fc containing other covariates that have effect on gene expression (based on variance plots)
# fc: MEDIAN_5PRIME_TO_3PRIME_BIAS, site_specimen_collected, PCT_MRNA_BASES, PCT_R1_TRANSCRIPT_STRAND_READS, rin, prep
# cb: PCT_MRNA_BASES, prep, site_specimen_collected, MEDIAN_3PRIME_BIAS, rin, PCT_INTERGENIC_BASES

#preparing DESeq dataset
dds2_fc <- DESeqDataSetFromMatrix(countData = fcCounts,
                              colData = fc_samples,
                              design = ~ MEDIAN_5PRIME_TO_3PRIME_BIAS + site_specimen_collected + PCT_MRNA_BASES + PCT_R1_TRANSCRIPT_STRAND_READS + rin + prep + disease_short)

#run differential analysis
dds2_fc <- DESeq(dds2_fc)

#Extract results to compare each disease group to control
fc_res2_als_vs_control <- results(dds2_fc, contrast = c("disease_short", "ALS", "Control"))
fc_res2_ftd_vs_control <- results(dds2_fc, contrast = c("disease_short", "FTD", "Control"))
fc_res2_als_ftd_vs_control <- results(dds2_fc, contrast = c("disease_short", "ALS/FTD", "Control"))
#fc_res2_als_other_vs_control <- results(dds2_fc, contrast = c("disease_short", "ALS/other", "Control"))
#fc_res2_other_vs_control <- results(dds2_fc, contrast = c("disease_short", "Other", "Control"))
fc_res2_als_ad_vs_control <- results(dds2_fc, contrast = c("disease_short", "ALS/AD", "Control"))

```

```{r}
#Control vs ALS (fc with multiple factors)
fc_res2_als_vs_control<- fc_res2_als_vs_control[!is.na(fc_res2_als_vs_control$padj), ]
fc_res2_als_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```

```{r}
#Control vs FTD (fc with multiple factors)
fc_res2_ftd_vs_control<- fc_res2_ftd_vs_control[!is.na(fc_res2_ftd_vs_control$padj), ]
fc_res2_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs ALS/FTD (fc with multiple factors)
fc_res2_als_ftd_vs_control<- fc_res2_als_ftd_vs_control[!is.na(fc_res2_als_ftd_vs_control$padj), ]
fc_res2_als_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```

```{r}
# #Control vs ALS/other (fc with multiple factors)
# fc_res2_als_other_vs_control<- fc_res2_als_other_vs_control[!is.na(fc_res2_als_other_vs_control$padj), ]
# fc_res2_als_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
# #Control vs Other (fc with multiple factors)
# fc_res2_other_vs_control<- fc_res2_other_vs_control[!is.na(fc_res2_other_vs_control$padj), ]
# fc_res2_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs ALS/AD (fc with multiple factors)
fc_res2_als_ad_vs_control<- fc_res2_als_ad_vs_control[!is.na(fc_res2_als_ad_vs_control$padj), ]
fc_res2_als_ad_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```


```{r}
#fc summary table for model with several factors
fc_degs2_als_vs_control <- fc_res2_als_vs_control[which(fc_res2_als_vs_control$padj < 0.05), ]
fc_degs2_ftd_vs_control <- fc_res2_ftd_vs_control[which(fc_res2_ftd_vs_control$padj < 0.05), ]
fc_degs2_als_ftd_vs_control <- fc_res2_als_ftd_vs_control[which(fc_res2_als_ftd_vs_control$padj < 0.05), ]
#fc_degs2_als_other_vs_control <- fc_res2_als_other_vs_control[which(fc_res2_als_other_vs_control$padj < 0.05), ]
#fc_degs2_other_vs_control <- fc_res2_other_vs_control[which(fc_res2_other_vs_control$padj < 0.05), ]
fc_degs2_als_ad_vs_control <- fc_res2_als_ad_vs_control[which(fc_res2_als_ad_vs_control$padj < 0.05), ]

# summary_table2_fc <- data.frame(
#   Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/other vs Control", "Other vs Control", "ALS/AD vs Control"),
#   Number_of_DEGs = c(nrow(fc_degs2_als_vs_control), nrow(fc_degs2_ftd_vs_control), nrow(fc_degs2_als_ftd_vs_control), nrow(fc_degs2_als_other_vs_control), nrow(fc_degs2_other_vs_control), nrow(fc_degs2_als_ad_vs_control)))

summary_table2_fc <- data.frame(
  Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/AD vs Control"),
  Number_of_DEGs = c(nrow(fc_degs2_als_vs_control), nrow(fc_degs2_ftd_vs_control), nrow(fc_degs2_als_ftd_vs_control), nrow(fc_degs2_als_ad_vs_control)))
summary_table2_fc

```

```{r}
#creating second model containing other covariates that have effect on gene expression (based on variance plots)
# cb: PCT_MRNA_BASES, prep, site_specimen_collected, MEDIAN_3PRIME_BIAS, rin, PCT_INTERGENIC_BASES

#preparing DESeq dataset
dds2_cb <- DESeqDataSetFromMatrix(countData = cbCounts,
                              colData = cb_samples,
                              design = ~ PCT_MRNA_BASES + prep + site_specimen_collected + MEDIAN_3PRIME_BIAS + rin + PCT_INTERGENIC_BASES + disease_short)

#run differential analysis
dds2_cb <- DESeq(dds2_cb)

#Extract results to compare each disease group to control
cb_res2_als_vs_control <- results(dds2_cb, contrast = c("disease_short", "ALS", "Control"))
cb_res2_ftd_vs_control <- results(dds2_cb, contrast = c("disease_short", "FTD", "Control"))
cb_res2_als_ftd_vs_control <- results(dds2_cb, contrast = c("disease_short", "ALS/FTD", "Control"))
#cb_res2_als_other_vs_control <- results(dds2_cb, contrast = c("disease_short", "ALS/other", "Control"))
#cb_res2_other_vs_control <- results(dds2_cb, contrast = c("disease_short", "Other", "Control"))
cb_res2_als_ad_vs_control <- results(dds2_cb, contrast = c("disease_short", "ALS/AD", "Control"))


#Control vs ALS (cb with multiple factors)
cb_res2_als_vs_control<- cb_res2_als_vs_control[!is.na(cb_res2_als_vs_control$padj), ]
cb_res2_als_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```

```{r}
#Control vs FTD (cb with multiple factors)
cb_res2_ftd_vs_control<- cb_res2_ftd_vs_control[!is.na(cb_res2_ftd_vs_control$padj), ]
cb_res2_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```


```{r}
#Control vs ALS/FTD (cb with multiple factors)
cb_res2_als_ftd_vs_control<- cb_res2_als_ftd_vs_control[!is.na(cb_res2_als_ftd_vs_control$padj), ]
cb_res2_als_ftd_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs ALS/other (cb with multiple factors)
# cb_res2_als_other_vs_control<- cb_res2_als_other_vs_control[!is.na(cb_res2_als_other_vs_control$padj), ]
# cb_res2_als_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs Other (cb with multiple factors)
# cb_res2_other_vs_control<- cb_res2_other_vs_control[!is.na(cb_res2_other_vs_control$padj), ]
# cb_res2_other_vs_control %>%
#   ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
#   geom_point(aes(color = padj <0.05))
```

```{r}
#Control vs ALS/AD (cb with multiple factors)
cb_res2_als_ad_vs_control<- cb_res2_als_ad_vs_control[!is.na(cb_res2_als_ad_vs_control$padj), ]
cb_res2_als_ad_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```

```{r}
#cb summary table for second model containing multiple factors
cb_degs2_als_vs_control <- cb_res2_als_vs_control[which(cb_res2_als_vs_control$padj < 0.05), ]
cb_degs2_ftd_vs_control <- cb_res2_ftd_vs_control[which(cb_res2_ftd_vs_control$padj < 0.05), ]
cb_degs2_als_ftd_vs_control <- cb_res2_als_ftd_vs_control[which(cb_res2_als_ftd_vs_control$padj < 0.05), ]
#cb_degs2_als_other_vs_control <- cb_res2_als_other_vs_control[which(cb_res2_als_other_vs_control$padj < 0.05), ]
#cb_degs2_other_vs_control <- cb_res2_other_vs_control[which(cb_res2_other_vs_control$padj < 0.05), ]
cb_degs2_als_ad_vs_control <- cb_res2_als_ad_vs_control[which(cb_res2_als_ad_vs_control$padj < 0.05), ]

summary_table2_cb <- data.frame(
  Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/AD vs Control"),
  Number_of_DEGs = c(nrow(cb_degs2_als_vs_control), nrow(cb_degs2_ftd_vs_control), nrow(cb_degs2_als_ftd_vs_control), nrow(cb_degs2_als_ad_vs_control)))
summary_table2_cb
```


```{r}
#creating third model for fc containing different combination of covariates that have effect on gene expression (based on variance plots)
# fc: MEDIAN_5PRIME_TO_3PRIME_BIAS, site_specimen_collected, PCT_MRNA_BASES, disease_short

#preparing DESeq dataset
dds3_fc <- DESeqDataSetFromMatrix(countData = fcCounts,
                              colData = fc_samples,
                              design = ~ MEDIAN_5PRIME_TO_3PRIME_BIAS + site_specimen_collected + PCT_MRNA_BASES + disease_short)

#run differential analysis
dds3_fc <- DESeq(dds3_fc)

#Extract results to compare each disease group to control
fc_res3_als_vs_control <- results(dds3_fc, contrast = c("disease_short", "ALS", "Control"))
fc_res3_ftd_vs_control <- results(dds3_fc, contrast = c("disease_short", "FTD", "Control"))
fc_res3_als_ftd_vs_control <- results(dds3_fc, contrast = c("disease_short", "ALS/FTD", "Control"))
#fc_res3_als_other_vs_control <- results(dds3_fc, contrast = c("disease_short", "ALS/other", "Control"))
#fc_res3_other_vs_control <- results(dds3_fc, contrast = c("disease_short", "Other", "Control"))
fc_res3_als_ad_vs_control <- results(dds3_fc, contrast = c("disease_short", "ALS/AD", "Control"))
```

```{r}
#fc summary table for third model containing top 3 factors with highest variance
fc_degs3_als_vs_control <- fc_res3_als_vs_control[which(fc_res3_als_vs_control$padj < 0.05), ]
fc_degs3_ftd_vs_control <- fc_res3_ftd_vs_control[which(fc_res3_ftd_vs_control$padj < 0.05), ]
fc_degs3_als_ftd_vs_control <- fc_res3_als_ftd_vs_control[which(fc_res3_als_ftd_vs_control$padj < 0.05), ]
#fc_degs3_als_other_vs_control <- fc_res3_als_other_vs_control[which(fc_res3_als_other_vs_control$padj < 0.05), ]
#fc_degs3_other_vs_control <- fc_res3_other_vs_control[which(fc_res3_other_vs_control$padj < 0.05), ]
fc_degs3_als_ad_vs_control <- fc_res3_als_ad_vs_control[which(fc_res3_als_ad_vs_control$padj < 0.05), ]

summary_table3_fc <- data.frame(
  Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/AD vs Control"),
  Number_of_DEGs = c(nrow(fc_degs3_als_vs_control), nrow(fc_degs3_ftd_vs_control), nrow(fc_degs3_als_ftd_vs_control), nrow(fc_degs3_als_ad_vs_control)))
summary_table3_fc
```

```{r}
#creating third model for cb containing different combination of covariates that have effect on gene expression (based on variance plots)
# cb: PCT_MRNA_BASES, prep, site_specimen_collected, MEDIAN_3PRIME_BIAS, disease_short

#preparing DESeq dataset
dds3_cb <- DESeqDataSetFromMatrix(countData = cbCounts,
                              colData = cb_samples,
                              design = ~ PCT_MRNA_BASES + prep + site_specimen_collected + MEDIAN_3PRIME_BIAS + disease_short)

#run differential analysis
dds3_cb <- DESeq(dds3_cb)

#Extract results to compare each disease group to control
cb_res3_als_vs_control <- results(dds3_cb, contrast = c("disease_short", "ALS", "Control"))
cb_res3_ftd_vs_control <- results(dds3_cb, contrast = c("disease_short", "FTD", "Control"))
cb_res3_als_ftd_vs_control <- results(dds3_cb, contrast = c("disease_short", "ALS/FTD", "Control"))
#cb_res3_als_other_vs_control <- results(dds3_cb, contrast = c("disease_short", "ALS/other", "Control"))
#cb_res3_other_vs_control <- results(dds3_cb, contrast = c("disease_short", "Other", "Control"))
cb_res3_als_ad_vs_control <- results(dds3_cb, contrast = c("disease_short", "ALS/AD", "Control"))


#Control vs ALS (cb with multiple factors)
cb_res3_als_vs_control<- cb_res3_als_vs_control[!is.na(cb_res3_als_vs_control$padj), ]
cb_res3_als_vs_control %>%
  ggplot(aes (x = log2FoldChange, y = -log10(pvalue))) + 
  geom_point(aes(color = padj <0.05))

```


```{r}
#cb summary table for third model containing top 4 covariates with highest variance
cb_degs3_als_vs_control <- cb_res3_als_vs_control[which(cb_res3_als_vs_control$padj < 0.05), ]
cb_degs3_ftd_vs_control <- cb_res3_ftd_vs_control[which(cb_res3_ftd_vs_control$padj < 0.05), ]
cb_degs3_als_ftd_vs_control <- cb_res3_als_ftd_vs_control[which(cb_res3_als_ftd_vs_control$padj < 0.05), ]
#cb_degs3_als_other_vs_control <- cb_res3_als_other_vs_control[which(cb_res3_als_other_vs_control$padj < 0.05), ]
#cb_degs3_other_vs_control <- cb_res3_other_vs_control[which(cb_res3_other_vs_control$padj < 0.05), ]
cb_degs3_als_ad_vs_control <- cb_res3_als_ad_vs_control[which(cb_res3_als_ad_vs_control$padj < 0.05), ]

summary_table3_cb <- data.frame(
  Comparison = c("ALS vs Control", "FTD vs Control", "ALS/FTD vs Control", "ALS/AD vs Control"),
  Number_of_DEGs = c(nrow(cb_degs3_als_vs_control), nrow(cb_degs3_ftd_vs_control), nrow(cb_degs3_als_ftd_vs_control), nrow(cb_degs3_als_ad_vs_control)))
summary_table3_cb

```


```{r}
#save fc_samples, fcCounts
save(fc_samples, fcCounts, file = "fc_samples.RData")
#save fc differential analysis result (model with factors)
save(dds2_fc, fc_res2_als_vs_control, fc_res2_ftd_vs_control, fc_res2_als_ftd_vs_control, fc_res2_als_ad_vs_control, file = "fc_deseq2.RData")
#save cb_samples, cbCounts
save(cb_samples, cbCounts, file = "cb_samples.RData")
#save cb differential analysis result (model with factors)
save(dds2_cb, cb_res2_als_vs_control, cb_res2_ftd_vs_control, cb_res2_als_ftd_vs_control, cb_res2_als_ad_vs_control, file = "cb_deseq2.RData")

save(pca_cb_sva, file = "cb_pca.RData")
```



```{r}
library(limma)
library(sva)
library(edgeR)
library(ggfortify)

# Use sva (surrogate variable analysis) to capture unknown variation
# Specify variables to be included in the voom() estimates of uncertainty.
# Recommend including variables with a small number of categories that explain a substantial amount of variation
#design <- model.matrix( ~ factor(site) + factor(prep), support)
#age_at_death?
design <- model.matrix(~factor(disease_short) + factor(site_specimen_collected) + factor(prep) + rin, data=fc_samples)
gExpr <- DGEList(fcCounts)
gExpr <- calcNormFactors(gExpr,method = "TMM")
fcCounts_voom <- voom(gExpr)$E

fc_samples$prep <- as.numeric(fc_samples$prep)
fc_samples$disease_short <- as.factor(as.character(fc_samples$disease_short))
fc_samples$site_specimen_collected <- as.factor(fc_samples$site_specimen_collected)


#corrected_expr <- removeBatchEffect(fcCounts, batch=known_tech_variables, covariates=sva_results$sv)
#batch -> categorical
#covariates are continuous


#technical factors in fc that are causing higher variances in expression
#covariate_df <- dplyr:: select(fc_samples, MEDIAN_5PRIME_TO_3PRIME_BIAS, PCT_MRNA_BASES, PCT_R1_TRANSCRIPT_STRAND_READS, rin, prep)
#removed MEDIAN_5PRIME_TO_3PRIME_BIAS and rin from list because of the big difference between ALS and FTD disease when looking at the effect of rin and MEDIAN_5PRIME_TO_3PRIME_BIAS across disease groups --> these factors may have signficant impact on differences between disease groups ALS and FTD
covariate_df <- dplyr:: select(fc_samples, PCT_MRNA_BASES, PCT_R1_TRANSCRIPT_STRAND_READS, prep)

covariate_df$prep <- as.numeric(covariate_df$prep)

counts_rbe <- (removeBatchEffect(fcCounts_voom,
                                       #batch = fc_samples$site_specimen_collected,
                                       covariates = covariate_df))

#head(counts_rbe)

rownames(fc_samples) <- fc_samples$external_sample_id

```


```{r}
pca_rbe <- prcomp(t(counts_rbe), scale = TRUE, center = TRUE)
autoplot(pca_rbe, color = "disease_short", data = fc_samples)
```

 
```{r}
pca_voom <- prcomp(t(fcCounts_voom), scale = TRUE, center = TRUE)
autoplot(pca_voom, color = "disease_short", data = fc_samples)
```




```{r}
library(patchwork)
#matrix
pca_df <- pca_rbe$x %>% as.data.frame() %>%
 rownames_to_column("external_sample_id") %>%
  left_join(fc_samples, by= "external_sample_id") 

ggplot(pca_df, aes(x=disease_short, y= PC1)) +
  geom_boxplot() +

ggplot(pca_df, aes(x=disease_short, y= PC2)) +
  geom_boxplot() +
  
  ggplot(pca_df, aes(x=disease_short, y= PC3)) +
  geom_boxplot()

```

```{r}
plot(pca_rbe)
```

```{r}
plot(pca_voom)
```


```{r}
pca_df_voom <- pca_voom$x %>% as.data.frame() %>%
   rownames_to_column("external_sample_id") %>%
  left_join(fc_samples, by= "external_sample_id") 

ggplot(pca_df_voom, aes(x=disease_short, y= PC1)) +
  geom_boxplot() +
  
ggplot(pca_df_voom, aes(x=disease_short, y= PC2)) +
  geom_boxplot() +

ggplot(pca_df_voom, aes(x=disease_short, y= PC3)) +
  geom_boxplot() 
```


```{r}
# Create model matrices
#known technical variables for fc: MEDIAN_5PRIME_TO_3PRIME_BIAS, PCT_MRNA_BASES, PCT_R1_TRANSCRIPT_STRAND_READS + rin + prep
mod0 <- model.matrix(~PCT_MRNA_BASES+ PCT_R1_TRANSCRIPT_STRAND_READS + prep, data=fc_samples)  # Model without disease condition (null model)
# Full model includes disease (protected) and known technical covariates
mod <- model.matrix(~ disease_short + PCT_MRNA_BASES+ PCT_R1_TRANSCRIPT_STRAND_READS + prep, data=fc_samples)  # Model with biological variable of interest (disease condition)

# Estimate surrogate variables (SVs)
# sva_results contains SVs, unobserved confounders estimated from the expression matrix.
num_sv <-num.sv(fcCounts_voom, mod, method = "be")
#sva_results <- sva(fcCounts, mod, mod0)

#object containing the sv variables
sv_obj <- sva(fcCounts_voom, mod, mod0, n.sv = num_sv)

# Regress out SVs from the voom matrix
voom_corrected <- removeBatchEffect(fcCounts_voom, covariates = sv_obj$sv)

#fcCounts_voom_sva <- sva_network(fcCounts_voom, num_sv)

pca_sva <- prcomp(t(voom_corrected), center = TRUE, scale = TRUE)

plot(pca_sva)
```

```{r}
autoplot(pca_sva, color = "disease_short", data = fc_samples)

```

```{r}
#Finding genes associated with pc1 and seeing if it aligns with differentially expressed genes found from deSeq

pc1_loadings_fc <- pca_sva$rotation[, 1]  # PC1 loadings

# Top 10 genes (positively corr with PC1) 
top_pc1_genes_fc <- sort(pc1_loadings_fc, decreasing = TRUE)[1:10]
print(top_pc1_genes_fc)
#ENSG00000186350.11 ENSG00000196576.15 ENSG00000178209.15 ENSG00000172081.14 ENSG00000105397.13 ENSG00000168056.15  ENSG00000173548.8 ENSG00000187535.14 ENSG00000165458.14 ENSG00000186174.12

# Top 10 genes (negatively correlated with PC1)
bottom_pc1_genes_fc <- sort(pc1_loadings_fc, decreasing = FALSE)[1:10]
print(bottom_pc1_genes_fc)
#ENSG00000198836.9 ENSG00000198961.10 ENSG00000104381.12 ENSG00000131507.11 ENSG00000129636.12 ENSG00000050748.17 ENSG00000131437.15 ENSG00000175395.16 ENSG00000136531.16 ENSG00000104093.13 

#top 10 genes (disregarding positive or negative corr) with PC1
top_pc1_genes_fc <- head(order(abs(pca_sva$rotation[,1]), decreasing = TRUE), 10)
top_pc1_gene_names_fc <- rownames(pca_sva$rotation)[top_pc1_genes_fc]
print(top_pc1_genes_fc) #row in PC
print(top_pc1_gene_names_fc) #corresponding name of gene

#deseq model of fc with technical variables included: design = ~ PCT_MRNA_BASES + prep + site_specimen_collected + MEDIAN_3PRIME_BIAS + rin + PCT_INTERGENIC_BASES + disease_short
#dds2_fc <- DESeq(dds2_fc) 
res_dds2_fc <- results(dds2_fc)

#order by padj and subset it for padj values < 0.05
res_padj_fc <- res_dds2_fc[order(res_dds2_fc$padj), ]
res_sig_fc <- subset(res_padj_fc, padj < 0.05)
fc_sig_degs <- rownames(res_sig_fc)

#intersect top pc1 gene names in fc and significant degs (based on deseq model with technical variables)
fc_sig_degs_intersect <- intersect(fc_sig_degs, top_pc1_gene_names_fc)
fc_sig_degs_intersect
# "ENSG00000198836.9"  "ENSG00000175395.16" "ENSG00000050748.17" "ENSG00000129636.12" "ENSG00000104093.13" "ENSG00000198961.10" "ENSG00000104381.12" "ENSG00000131437.15" "ENSG00000131507.11" "ENSG00000136531.16"

```















```{r}
#plotting top 5 pcs of fc against log fold change for FTD_vs_Control

#create pc1 dataframe and name the columns gene_id and PC1
pc1_fc_df <- enframe(pc1_loadings_fc, name = "gene_id", value = "PC1")

pc2_loadings_fc <- pca_sva$rotation[, 2] 
pc2_fc_df <- enframe(pc2_loadings_fc, name = "gene_id", value = "PC2")

pc3_loadings_fc <- pca_sva$rotation[, 3] 
pc3_fc_df <- enframe(pc3_loadings_fc, name = "gene_id", value = "PC3")

pc4_loadings_fc <- pca_sva$rotation[, 4] 
pc4_fc_df <- enframe(pc4_loadings_fc, name = "gene_id", value = "PC4")

pc5_loadings_fc <- pca_sva$rotation[, 5] 
pc5_fc_df <- enframe(pc5_loadings_fc, name = "gene_id", value = "PC5")


#change row names to column for results of FC's FTD vs Control results and add under gene_id column
fc_res2_ftd_vs_control <- rownames_to_column(as.data.frame(fc_res2_ftd_vs_control), "gene_id" )

#left join pc1 dataframe and FTD vs Control results dataframe by the gene id 
pc_fc_FTD_df <- left_join(pc1_fc_df, fc_res2_ftd_vs_control, by = "gene_id")
pc_fc_FTD_df <- left_join(pc2_fc_df, pc_fc_FTD_df, by = "gene_id")
pc_fc_FTD_df <- left_join(pc3_fc_df, pc_fc_FTD_df, by = "gene_id")
pc_fc_FTD_df <- left_join(pc4_fc_df, pc_fc_FTD_df, by = "gene_id")
pc_fc_FTD_df <- left_join(pc5_fc_df, pc_fc_FTD_df, by = "gene_id")


ggplot(pc_fc_FTD_df, aes(x =PC1, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_fc_FTD_df, aes(x =PC2, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_fc_FTD_df, aes(x =PC3, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_fc_FTD_df, aes(x=PC4, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_fc_FTD_df, aes(x =PC5, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) 

pc_fc_FTD_df %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05") +
  ggpubr::stat_cor()

```

```{r}
#change row names to column for results of FC's ALS vs Control results and add under gene_id column
fc_res2_als_vs_control <- rownames_to_column(as.data.frame(fc_res2_als_vs_control), "gene_id")

#create combined data frame for PCs 1-5 and DEGs in ALS vs Control result in fc
pc_fc_ALS_df <- left_join(pc1_fc_df, fc_res2_als_vs_control, by = "gene_id")
pc_fc_ALS_df <- left_join(pc2_fc_df, pc_fc_ALS_df, by = "gene_id")
pc_fc_ALS_df <- left_join(pc3_fc_df, pc_fc_ALS_df, by = "gene_id")
pc_fc_ALS_df <- left_join(pc4_fc_df, pc_fc_ALS_df)
pc_fc_ALS_df <- left_join(pc5_fc_df, pc_fc_ALS_df)


pc_fc_ALS_df %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05") +
  ggpubr::stat_cor()

```

```{r}
#change row names to column for results of FC's ALS/FTD vs Control results and add under gene_id column
fc_res2_als_ftd_vs_control <- rownames_to_column(as.data.frame(fc_res2_als_ftd_vs_control), "gene_id")

#create combined data frame for PCs 1-5 and DEGs in ALS/FTD vs Control result in fc
pc_fc_ALS_FTD_df <- left_join(pc1_fc_df, fc_res2_als_ftd_vs_control, by = "gene_id")
pc_fc_ALS_FTD_df <- left_join(pc2_fc_df, pc_fc_ALS_FTD_df, by = "gene_id")
pc_fc_ALS_FTD_df <- left_join(pc3_fc_df, pc_fc_ALS_FTD_df, by = "gene_id")
pc_fc_ALS_FTD_df <- left_join(pc4_fc_df, pc_fc_ALS_FTD_df, by = "gene_id")
pc_fc_ALS_FTD_df <- left_join(pc5_fc_df, pc_fc_ALS_FTD_df, by = "gene_id")


pc_fc_ALS_FTD_df %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05")

```

```{r}
#change row names to column for results of FC's ALS/AD vs Control results and add under gene_id column
fc_res2_als_ad_vs_control <- rownames_to_column(as.data.frame(fc_res2_als_ad_vs_control), "gene_id")

#create combined data frame for PCs 1-5 and DEGs in ALS/AD vs Control result in fc
pc_fc_ALS_AD_df <- left_join(pc1_fc_df, fc_res2_als_ad_vs_control, by = "gene_id")
pc_fc_ALS_AD_df <- left_join(pc2_fc_df, pc_fc_ALS_AD_df, by = "gene_id")
pc_fc_ALS_AD_df <- left_join(pc3_fc_df, pc_fc_ALS_AD_df, by = "gene_id")
pc_fc_ALS_AD_df <- left_join(pc4_fc_df, pc_fc_ALS_AD_df, by = "gene_id")
pc_fc_ALS_AD_df <- left_join(pc5_fc_df, pc_fc_ALS_AD_df, by = "gene_id")


pc_fc_ALS_AD_df %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05")

```

```{r}
pc2_loadings_fc <- pca_sva$rotation[, 2] 

pc2_fc_df <- enframe(pc2_loadings_fc, name = "gene_id", value = "PC2")

pc2_fc_df <- left_join(pc2_fc_df, fc_res2_ftd_vs_control, by = "gene_id")

pc2_fc_df %>%
ggplot(aes(x =PC2, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5)

```

top 5 pcs 
correlate PCs against log2foldchange
data frame containing 
ggplot with three columns (comparison, PC, foldchange, PC_loading, padj)
scatterplot with PC against fold change
facet_wrap/grid by comparison + PC

for each disease comparison

```{r}
library(patchwork)
pca_df_sva<- pca_sva$x %>% as.data.frame() %>%
   rownames_to_column("external_sample_id") %>%
  left_join(fc_samples, by= "external_sample_id") 

ggplot(pca_df_sva, aes(x=disease_short, y= PC1)) +
  geom_boxplot() +
  
ggplot(pca_df_voom, aes(x=disease_short, y= PC2)) +
  geom_boxplot() +

ggplot(pca_df_voom, aes(x=disease_short, y= PC3)) +
  geom_boxplot() 

```

```{r}
library(umap)
#autoplot(pca_sva, color = "disease_short", data = fc_samples)

umap_fc_sva <- umap(pca_sva$x[,1:5])
umap_fc_df <- as.data.frame(umap_fc_sva$layout)
#setting names of data frame
names(umap_fc_df) <- c("umap1", "umap2") 
#rownames to column
rownames_to_column(umap_fc_df, "external_sample_id") %>%
  left_join(fc_samples, by = "external_sample_id") %>%
  ggplot(aes(x=umap1, y=umap2, color= disease_short)) +
  geom_point()

```



```{r}
# 
# library(RUVSeq)
# 
# # Create SeqExpressionSet
# set <- newSeqExpressionSet(as.matrix(fcCounts_voom),
#                            phenoData = data.frame(fc_samples, row.names=colnames(fcCounts_voom)))
# 
# # Use RUVr (e.g., if you don’t have controls or replicates)
# design <- model.matrix(~ disease_short, data = fc_samples)
# fit <- edgeR::glmFit(edgeR::DGEList(counts), design)
# 
# residuals <- residuals(fit, type = "deviance")
# 
# set_ruv <- RUVr(set, cIdx = rownames(counts), k = 5, residuals)
# 
# ruv_corrected_counts <- normCounts(set_ruv)

```


```{r}
#cerebellum - regressing out variations
library(limma)
library(sva)
library(edgeR)
library(ggfortify)

#technical factors in cb that are causing higher variances in expression: PCT_MRNA_BASES + prep + site_specimen_collected + MEDIAN_3PRIME_BIAS + rin + PCT_INTERGENIC_BASES + disease_short
#design <- model.matrix(~factor(disease_short) + factor(site_specimen_collected) + factor(prep) + rin, data=cb_samples)
gExpr <- DGEList(cbCounts)
gExpr <- calcNormFactors(gExpr,method = "TMM")
cbCounts_voom <- voom(gExpr)$E

cb_samples$prep <- as.numeric(cb_samples$prep)
cb_samples$disease_short <- as.factor(as.character(cb_samples$disease_short))
cb_samples$site_specimen_collected <- as.factor(cb_samples$site_specimen_collected)

#corrected_expr <- removeBatchEffect(fcCounts, batch=known_tech_variables, covariates=sva_results$sv)
#batch -> categorical
#covariates are continuous

#covariate_df <- dplyr:: select(cb_samples, PCT_MRNA_BASES, prep, site_specimen_collected, MEDIAN_3PRIME_BIAS, rin, PCT_INTERGENIC_BASES, disease_short)
#removed rin from list because of the big difference between ALS and FTD disease when looking at the effect of rin and MEDIAN_5PRIME_TO_3PRIME_BIAS across disease groups --> these factors may have signficant impact on differences between disease groups ALS and FTD
covariate_cb_df <- dplyr:: select(cb_samples, PCT_MRNA_BASES, prep, MEDIAN_3PRIME_BIAS, rin, PCT_INTERGENIC_BASES)

covariate_cb_df$prep <- as.numeric(covariate_cb_df$prep)

#13684 sites are NA
cbCounts_rbe <- (removeBatchEffect(cbCounts_voom,
                                       #batch = cb_samples$site_specimen_collected,
                                       covariates = covariate_cb_df))

#head(cbCounts_rbe)

rownames(cb_samples) <- cb_samples$external_sample_id

```

```{r}
pca_cb_rbe <- prcomp(t(cbCounts_rbe), scale = TRUE, center = TRUE)
autoplot(pca_cb_rbe, color = "disease_short", data = cb_samples)
```
 
```{r}
plot(pca_cb_rbe)
```


```{r}
library(patchwork)
#matrix
pca_cb_df <- pca_cb_rbe$x %>% as.data.frame() %>%
 rownames_to_column("external_sample_id") %>%
  left_join(cb_samples, by= "external_sample_id") 

ggplot(pca_cb_df, aes(x=disease_short, y= PC1)) +
  geom_boxplot() +

ggplot(pca_cb_df, aes(x=disease_short, y= PC2)) +
  geom_boxplot() +
  
  ggplot(pca_cb_df, aes(x=disease_short, y= PC3)) +
  geom_boxplot()

```

```{r}
pca_cb_voom <- prcomp(t(cbCounts_voom), scale = TRUE, center = TRUE)
autoplot(pca_cb_voom, color = "disease_short", data = cb_samples)
```

```{r}
plot(pca_cb_voom)
```

```{r}
# Create model matrices
#known technical variables for cb: PCT_MRNA_BASES, PCT_R1_TRANSCRIPT_STRAND_READS + rin + prep
mod0_cb <- model.matrix(~rin + prep + MEDIAN_3PRIME_BIAS + PCT_INTERGENIC_BASES, data=cb_samples)  # Model without disease condition (null model)
#mod0_cb <- model.matrix(~1, data=cb_samples)
# Full model includes disease (protected) and known technical covariates
mod_cb <- model.matrix(~ disease_short + rin + prep + MEDIAN_3PRIME_BIAS + PCT_INTERGENIC_BASES, data=cb_samples)  # Model with biological variable of interest (disease condition)
#mod_cb <- model.matrix(~ disease_short, data=cb_samples) 

# Estimate surrogate variables (SVs)
# sva_results contains SVs, unobserved confounders estimated from the expression matrix.
num_sv_cb <-num.sv(cbCounts_voom, mod_cb, method = "be")
#sva_results <- sva(fcCounts, mod, mod0)

#object containing the sv variables
sv_obj_cb <- sva(cbCounts_voom, mod_cb, mod0_cb, n.sv = num_sv_cb)

#correlation matrix checking if sv variables are correlated to technical variables but not disease_short

# Regress out SVs from the voom matrix
voom_corrected_cb <- removeBatchEffect(cbCounts_voom, covariates = sv_obj_cb$sv)

#fcCounts_voom_sva <- sva_network(fcCounts_voom, num_sv)

pca_cb_sva <- prcomp(t(voom_corrected_cb), center = TRUE, scale = TRUE)

plot(pca_cb_sva)
```

```{r}
#cb sig degs
pc1_loadings_cb<- pca_cb_sva$rotation[, 1]  # PC1 loadings in cb

# Top 10 genes(positively corr with PC1) 
top_pc1_genes_cb <- sort(pc1_loadings_cb, decreasing = TRUE)[1:10]
print(top_pc1_genes_cb)
#ENSG00000172667.11 ENSG00000187231.14 ENSG00000164114.19 ENSG00000196586.14 ENSG00000152332.16 ENSG00000091039.17 ENSG00000014123.10  ENSG00000184203.8 ENSG00000085832.17 ENSG00000134352.20 

# Top 10 genes most negatively correlated with PC1
bottom_pc1_genes_cb <- sort(pc1_loadings_cb, decreasing = FALSE)[1:10]
print(bottom_pc1_genes_cb)
# ENSG00000128159.12 ENSG00000138834.12 ENSG00000167861.16 ENSG00000065717.15 ENSG00000154832.14 ENSG00000128482.16 ENSG00000198910.13 ENSG00000178950.17 ENSG00000137802.14 ENSG00000146830.10  

#top 10 genes (not regarding postivie or negative) with PC1
top_pc1_genes_cb <- head(order(abs(pca_cb_sva$rotation[,1]), decreasing = TRUE), 10)
top_pc1_gene_names_cb <- rownames(pca_cb_sva$rotation)[top_pc1_genes_cb]
print(top_pc1_genes_cb) #row in PC
print(top_pc1_gene_names_cb) #corresponding name of gene

#deseq model of cb with technical variables included:  design = ~ PCT_MRNA_BASES + prep + site_specimen_collected + MEDIAN_3PRIME_BIAS + rin + PCT_INTERGENIC_BASES + disease_short
#dds2_fc <- DESeq(dds2_fc) 
res_dds2_cb <- results(dds2_cb)

#order by padj and subset it for padj values < 0.05
res_padj_cb <- res_dds2_cb[order(res_dds2_cb$padj), ]
res_sig_cb <- subset(res_padj_cb, padj < 0.05)
cb_sig_degs <- rownames(res_sig_cb)

#intersect top pc1 gene names in fc and significant degs (based on deseq model with technical variables)
cb_sig_degs_intersect <- intersect(cb_sig_degs, top_pc1_gene_names_cb)
sort(cb_sig_degs_intersect, decreasing= TRUE)
# "ENSG00000154832.14" "ENSG00000167861.16" "ENSG00000178950.17" "ENSG00000128159.12" "ENSG00000198910.13" "ENSG00000137802.14" "ENSG00000128482.16" "ENSG00000065717.15"

```

```{r}
#plotting top 5 pcs of cb against log fold change for FTD_vs_Control

#create pc1 dataframe and name the columns gene_id and PC1
pc1_cb_df <- enframe(pc1_loadings_cb, name = "gene_id", value = "PC1")

pc2_loadings_cb <- pca_cb_sva$rotation[, 2] 
pc2_cb_df <- enframe(pc2_loadings_cb, name = "gene_id", value = "PC2")

pc3_loadings_cb <- pca_cb_sva$rotation[, 3] 
pc3_cb_df <- enframe(pc3_loadings_cb, name = "gene_id", value = "PC3")

pc4_loadings_cb <- pca_cb_sva$rotation[, 4] 
pc4_cb_df <- enframe(pc4_loadings_cb, name = "gene_id", value = "PC4")

pc5_loadings_cb <- pca_cb_sva$rotation[, 5] 
pc5_cb_df <- enframe(pc5_loadings_cb, name = "gene_id", value = "PC5")


#change row names to column for results of cb's FTD vs Control results and add under gene_id column
cb_res2_ftd_vs_control <- rownames_to_column(as.data.frame(cb_res2_ftd_vs_control), "gene_id" )

#left join pc1 dataframe and FTD vs Control results dataframe by the gene id 
pc_cb_FTD_df <- left_join(pc1_cb_df, cb_res2_ftd_vs_control, by = "gene_id")
pc_cb_FTD_df <- left_join(pc2_cb_df, pc_cb_FTD_df, by = "gene_id")
pc_cb_FTD_df <- left_join(pc3_cb_df, pc_cb_FTD_df, by = "gene_id")
pc_cb_FTD_df <- left_join(pc4_cb_df, pc_cb_FTD_df, by = "gene_id")
pc_cb_FTD_df <- left_join(pc5_cb_df, pc_cb_FTD_df, by = "gene_id")


ggplot(pc_cb_FTD_df, aes(x =PC1, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_cb_FTD_df, aes(x =PC2, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_cb_FTD_df, aes(x =PC3, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_cb_FTD_df, aes(x=PC4, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) + 
ggplot(pc_cb_FTD_df, aes(x =PC5, y=log2FoldChange, color = padj < 0.05)) + 
  geom_point(size = 0.5) 

pc_cb_FTD_df %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05") +
  ggpubr::stat_cor()


```

```{r}
#change row names to column for results of CB's ALS vs Control results and add under gene_id column
cb_res2_als_vs_control <- rownames_to_column(as.data.frame(cb_res2_als_vs_control), "gene_id")

#create combined data frame for PCs 1-5 and DEGs in ALS vs Control result in fc
pc_cb_ALS_df <- left_join(pc1_cb_df, cb_res2_als_vs_control, by = "gene_id")
pc_cb_ALS_df <- left_join(pc2_cb_df, pc_cb_ALS_df, by = "gene_id")
pc_cb_ALS_df <- left_join(pc3_cb_df, pc_cb_ALS_df, by = "gene_id")
pc_cb_ALS_df <- left_join(pc4_cb_df, pc_cb_ALS_df, by = "gene_id")
pc_cb_ALS_df <- left_join(pc5_cb_df, pc_cb_ALS_df, by = "gene_id")


pc_cb_ALS_df %>%
  filter(padj < 0.05) %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05") +
  ggpubr::stat_cor()

```

```{r}
#change row names to column for results of CB's ALS/FTD vs Control results and add under gene_id column
cb_res2_als_ftd_vs_control <- rownames_to_column(as.data.frame(cb_res2_als_ftd_vs_control), "gene_id")

#create combined data frame for PCs 1-5 and DEGs in ALS/FTD vs Control result in fc
pc_cb_ALS_FTD_df <- left_join(pc1_cb_df, cb_res2_als_ftd_vs_control, by = "gene_id")
pc_cb_ALS_FTD_df <- left_join(pc2_cb_df, pc_cb_ALS_FTD_df, by = "gene_id")
pc_cb_ALS_FTD_df <- left_join(pc3_cb_df, pc_cb_ALS_FTD_df, by = "gene_id")
pc_cb_ALS_FTD_df <- left_join(pc4_cb_df, pc_cb_ALS_FTD_df, by = "gene_id")
pc_cb_ALS_FTD_df <- left_join(pc5_cb_df, pc_cb_ALS_FTD_df, by = "gene_id")


pc_cb_ALS_FTD_df %>%
  filter(padj < 0.05) %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05") +
  ggpubr::stat_cor()
```

```{r}
#change row names to column for results of CB's ALS/AD vs Control results and add under gene_id column
cb_res2_als_ad_vs_control <- rownames_to_column(as.data.frame(cb_res2_als_ad_vs_control), "gene_id")

#create combined data frame for PCs 1-5 and DEGs in ALS/AD vs Control result in fc
pc_cb_ALS_AD_df <- left_join(pc1_cb_df, cb_res2_als_ad_vs_control, by = "gene_id")
pc_cb_ALS_AD_df <- left_join(pc2_cb_df, pc_cb_ALS_AD_df, by = "gene_id")
pc_cb_ALS_AD_df <- left_join(pc3_cb_df, pc_cb_ALS_AD_df, by = "gene_id")
pc_cb_ALS_AD_df <- left_join(pc4_cb_df, pc_cb_ALS_AD_df, by = "gene_id")
pc_cb_ALS_AD_df <- left_join(pc5_cb_df, pc_cb_ALS_AD_df, by = "gene_id")


pc_cb_ALS_AD_df %>%
  pivot_longer(cols = starts_with("PC"), names_to = "PC", values_to = "PC_value") %>%
  ggplot(aes(x = PC_value, y = log2FoldChange, color = padj < 0.05)) +
  geom_point(size = 0.5) +
  facet_wrap(~ PC, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Principal Component Value", y = "log2 Fold Change", color = "padj < 0.05")

```

```{r}
library(umap)
autoplot(pca_cb_sva, color = "disease_short", data = cb_samples)

umap_cb_sva <- umap(pca_cb_sva$x[,1:5])
umap_cb_df <- as.data.frame(umap_cb_sva$layout)
#setting names of data frame
names(umap_cb_df) <- c("umap1", "umap2") 
#rownames to column
rownames_to_column(umap_cb_df, "external_sample_id") %>%
  left_join(cb_samples, by = "external_sample_id") %>%
  ggplot(aes(x=umap1, y=umap2, color= disease_short)) +
  geom_point()
```


```{r}
library(patchwork)
pca_cb_df_sva<- pca_cb_sva$x %>% as.data.frame() %>%
   rownames_to_column("external_sample_id") %>%
  left_join(cb_samples, by= "external_sample_id") 

ggplot(pca_cb_df_sva, aes(x=disease_short, y= PC1)) +
  geom_boxplot() +
  
ggplot(pca_cb_df_sva, aes(x=disease_short, y= PC2)) +
  geom_boxplot() +

ggplot(pca_cb_df_sva, aes(x=disease_short, y= PC3)) +
  geom_boxplot() 

```

```{r}
#comparing log2 fold changes between the different tissues and disease groups 

#inner join on gene name between the results of fc and cb for ALS vs control 
merged_als_vs_control <- merge(as.data.frame(fc_res_als_vs_control), as.data.frame(cb_res_als_vs_control), by="row.names")
merged_als_vs_control

ggplot(merged_als_vs_control, aes(x = log2FoldChange.x, y = log2FoldChange.y, color = log2FoldChange.x)) +
  geom_point() 

```

```{r}
#creating volcano plots

res_als_vs_control$log_pval <- -log10(res_als_vs_control$pvalue)
res_als_vs_control$significant <- ifelse(res_als_vs_control$padj < 0.05 & abs(res_als_vs_control$log2FoldChange) > 1, "Significant", "Not Significant")

ggplot(res_als_vs_control, aes(x = log2FoldChange, y = log_pval, color = significant)) +
  geom_point(alpha = 0.7) +
  scale_color_manual(values = c("Significant" = "red", "Not Significant" = "gray")) +
  theme_minimal() +
  labs(title = "ALS vs Control", x = "Log2 Fold Change", y = "-Log10 P-value") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed")

```





marker analysis:
```{r}
load("/Users/judyliang/Downloads/kelley_markers.RData")
gene_names <- read_tsv("gencode.v30.gene_meta.tsv")

#take kelley data and make into data frame
#add in gene IDs from gene meta
#join that with PC loadings by gene ID
#plot distribution of pc loadings by cell types

```